File: lesson_3/app/src/main/java/uz/micro/star/lesson_3/MainActivity.java
package uz.micro.star.lesson_3;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.Collections;

public class MainActivity extends AppCompatActivity {
    private Button[][] buttons = new Button[5][5];
    private ArrayList<Integer> numbers;
    private Coordinate object;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        loadViews();
        loadNumbers();
        shuffleNumbers();
        loadDataToViews();
    }

    private void loadViews() {
        buttons = new Button[4][4];
        ViewGroup group = findViewById(R.id.group);
        for (int i = 0; i < group.getChildCount(); i++) {
            buttons[i / 4][i % 4] = (Button) group.getChildAt(i);
            buttons[i / 4][i % 4].setOnClickListener(this::click);
            buttons[i / 4][i % 4].setTag(new Coordinate(i / 4, i % 4));
        }
    }

    private void loadNumbers() {
        numbers = new ArrayList<>();
        for (int i = 1; i <= 15; i++) {
            numbers.add(i);
        }
    }

    private void shuffleNumbers() {
        Collections.shuffle(numbers);
    }

    @SuppressLint("SetTextI18n")
    private void loadDataToViews() {
        object = new Coordinate(3, 3);
        for (int i = 0; i < 16; i++) {
            buttons[i / 4][i % 4].setText(i < 15 ? numbers.get(i) + "" : "");
            buttons[i / 4][i % 4].setBackgroundResource(i < 15 ? R.drawable.button_background : R.drawable.button_background_empty);
        }
    }

    public void restart(View view) {
        shuffleNumbers();
        loadDataToViews();
    }

    private void click(View view) {
        Button button = (Button) view;
        Coordinate c = (Coordinate) button.getTag();
        int delX = Math.abs(c.getX() - object.getX());
        int delY = Math.abs(c.getY() - object.getY());
        if (delX + delY == 1) {
            Button objectButton = buttons[object.getX()][object.getY()];
            objectButton.setText(button.getText());
            button.setText("");
            objectButton.setBackgroundResource(R.drawable.button_background);
            button.setBackgroundResource(R.drawable.button_background_empty);
            object.update(c);
        }
    }

    public void finish(View view) {
        finish();
    }
}

File: Lesson_6/app/src/main/java/uz/micro/star/lesson_6/MainActivity.kt
package uz.micro.star.lesson_6

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import uz.micro.star.lesson_6.managers.TestManager
import uz.micro.star.lesson_6.models.QuestionData
import uz.micro.star.lesson_6.utils.toast

class MainActivity : AppCompatActivity() {
    private lateinit var questionTextView: TextView
    private val allVariationsViewGroup by lazy {
        ArrayList<ViewGroup>()
    }
    private var selectedVariationImageView: ImageView? = null
    private lateinit var nextBtn: Button
    private val questionsList by lazy { ArrayList<QuestionData>() }
    lateinit var testManager: TestManager
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        getAllQuestions()
        testManager = TestManager(questionsList)
        loadView()
        loadDataToView()
    }

    private fun loadDataToView() {
        uncheck()
        questionTextView.text = testManager.getQuestion()
        (allVariationsViewGroup[0].getChildAt(1) as TextView).text = testManager.getVariantA()
        allVariationsViewGroup[0].setOnClickListener {
            selectVariation(it)
        }
        (allVariationsViewGroup[1].getChildAt(1) as TextView).text = testManager.getVariantB()
        allVariationsViewGroup[1].setOnClickListener {
            selectVariation(it)
        }
        (allVariationsViewGroup[2].getChildAt(1) as TextView).text = testManager.getVariantC()
        allVariationsViewGroup[2].setOnClickListener {
            selectVariation(it)
        }
        (allVariationsViewGroup[3].getChildAt(1) as TextView).text = testManager.getVariantD()
        allVariationsViewGroup[3].setOnClickListener {
            selectVariation(it)
        }
    }

    fun selectVariation(view: View) {
        val group = view as ViewGroup
        val selectedVariant = group.getChildAt(0) as ImageView
        uncheck()
        selectedVariant.setImageResource(R.drawable.ic_radio_button_checked_black_24dp)
        selectedVariationImageView = selectedVariant
    }

    private fun uncheck() {
        if (selectedVariationImageView != null) {
            selectedVariationImageView?.setImageResource(R.drawable.ic_radio_button_unchecked_black_24dp)
        }
    }

    private fun loadView() {
        questionTextView = findViewById(R.id.question)
        val viewGroup = findViewById<LinearLayout>(R.id.group)
        for (i in 0 until viewGroup.childCount) {
            if (viewGroup.getChildAt(i) is LinearLayout) {
                allVariationsViewGroup.add(viewGroup.getChildAt(i) as LinearLayout)
            }
        }
        nextBtn = findViewById(R.id.nextBtn)
        nextBtn.setOnClickListener {
            next()
        }
    }

    fun getAllQuestions() {
        questionsList.add(
            QuestionData(
                "Poytaxtimiz ?", "Buxoro",
                "Navoiy", "Samarqand", "Toshkent", "Toshkent"
            )
        )
        questionsList.add(
            QuestionData(
                "Eng kichik davlat", "Xitoy",
                "Vatikan", "Russia", "Ukraina", "Vatikan"
            )
        )
        questionsList.add(
            QuestionData(
                "2x2 nechi", "5",
                "1", "4", "53", "4"
            )
        )
        questionsList.add(
            QuestionData(
                "Ummon bu nima", "Guruh",
                "Davlat", "Okean", "Bilmadim", "Bilmadim"
            )
        )
    }

    fun next() {
        if (selectedVariationImageView != null) {
            val viewGroup = selectedVariationImageView!!.parent as LinearLayout
            val textview = viewGroup.getChildAt(1) as TextView
            testManager.checkAnswer(textview.text.toString())
            if (testManager.hasNextQuestion()) {
                loadDataToView()
            } else {
                toast("${testManager.correctAnswerCount} " +
                        "${testManager.wrongAnswerCount} ${testManager.percent}%")
            }
        } else {
            toast("Please choose one of variations ;)")
        }
    }
}